packages/@storefront/flux-capacitor/src/flux-capacitor.ts

```diff
@@ -52,12 +74,60 @@ class FluxCapacitor extends Emitter {
     delete this.__config;
   }
 
+  initHistory({ build, parse, initialUrl, pushState, replaceState }: History) {
+    this.history = {
+      ...this.history,
+      build,
+      initialUrl,
+      pushState,
+      replaceState,
+      parse: (u) => Promise
+        .resolve(parse(u))
+        .catch((e) => console.warn('UrlService parse promise failed', e)),
+    };
+
+    this.updateHistory({ shouldFetch: true, url: initialUrl, method: this.history.noop });
+  }
+
+  // this function is here to support legacy implementations >.<"
   saveState(route: string) {
-    this.emit(Events.HISTORY_SAVE, { route, state: this.store.getState() });
+    this.pushState({ route });
+  }
+
+  pushState(urlState: Partial<Actions.Payload.History.ProtoState>) {
+    this.updateHistory({ shouldFetch: true, ...urlState, method: this.history.pushState });
+  }
+
+  replaceState(route: string, buildAndParse: boolean = false) {
+    buildAndParse
+      ? this.updateHistory({ shouldFetch: false, route, method: this.history.replaceState })
+      // tslint:disable-next-line max-line-length
+      : this.store.dispatch(this.actions.updateHistory({ shouldFetch: false, route, method: this.history.replaceState }));
+  }
+
+  updateHistory(urlState: Partial<Actions.Payload.History.State> & { method: Function }) {
+    const url = urlState.url || this.history.build(urlState.route, this.store.getState());
+
+    try {
+      this.history.parse(url).then((result) => {
+        const { route, request } = <ParsedUrl>result;
+        this.store.dispatch(
+          this.actions.updateHistory({
+            url,
+            route: urlState.route || route,
+            request: urlState.request || request,
+            shouldFetch: urlState.shouldFetch,
+            method: urlState.method
+          })
+        );
+      });
+    } catch (e) {
```

Comment:
You can't get here unless something in the `this.history.parse(url)` chain is undefined. `parse` returns a Promise, which has its own error handling.

Reply:
This was kept in to preserve old behaviour. This is the warning we get when you return a `route` that isn't a known `route` (eg, run presets w/ just `http://localhost:8080/`)